mlir/lib/Dialect/Rock/Transforms/AddAsyncWait.cpp

```
@@ -0,0 +1,516 @@
+                // firstReadParentOp comes before user, so firstReadOp comes
+                // first Don't update, keep current firstRead
+              } else {
+                // user comes before firstReadParentOp, so user comes first
+                firstRead = readOp;
+                firstReadOp = user;
+              }
+            }
+            // Check if userBlock is nested inside firstReadBlock's region
+            else if (Operation *userParentOp = userBlock->getParentOp()) {
+              if (userParentOp->getBlock() == firstReadBlock) {
+                // userBlock is nested in firstReadBlock, compare userParentOp
+                // with firstReadOp
+                if (userParentOp->isBeforeInBlock(firstReadOp)) {
+                  // userParentOp comes before firstReadOp, so user comes first
+                  firstRead = readOp;
+                  firstReadOp = user;
+                }
+                // Otherwise firstReadOp comes first, don't update
+              }
+            }
+            // If neither is nested in the other, don't update firstRead
+          }
+        }
+      }
+      // If this is a view-like operation, follow its result
+      else if (auto viewOp = dyn_cast<ViewLikeOpInterface>(user)) {
+        addToWorklist(viewOp->getResult(0));
+      }
+      // If this is an extract_multibuffer, follow its result
+      else if (auto extractOp = dyn_cast<rock::ExtractMultiBufferOp>(user)) {
+        addToWorklist(extractOp.getResult());
+      }
+      // If this is a transform operation, follow its result
+      else if (auto transformOp = dyn_cast<rock::TransformOp>(user)) {
+        addToWorklist(transformOp.getResult());
+      }
+    }
+  }
+
+  return firstRead;
+}
+
+/// Find the first use after a ThreadwiseReadIntoOp that writes to LDS.
+/// The function traces back the dest value to the alloc, then finds the first
+/// ThreadwiseReadIntoOp that reads from that alloc.
+static Operation *
```

FailureOr<> instead of returning null